utils/duration_extraction.py

```diff
@@ -0,0 +1,95 @@
+import numpy as np
+from scipy.sparse import coo_matrix
+from scipy.sparse.csgraph import dijkstra
+
+
+def to_node_index(i, j, cols):
+    return cols * i + j
+
+
+def from_node_index(node_index, cols):
+    return node_index // cols, node_index % cols
+
+
+def to_adj_matrix(mat):
+    rows = mat.shape[0]
+    cols = mat.shape[1]
+
+    row_ind = []
+    col_ind = []
+    data = []
+
+    for i in range(rows):
+        for j in range(cols):
+
+            node = to_node_index(i, j, cols)
+
+            if j < cols - 1:
+                right_node = to_node_index(i, j + 1, cols)
+                weight_right = mat[i, j + 1]
+                row_ind.append(node)
+                col_ind.append(right_node)
+                data.append(weight_right)
+
+            if i < rows - 1 and j < cols:
+                bottom_node = to_node_index(i + 1, j, cols)
+                weight_bottom = mat[i + 1, j]
+                row_ind.append(node)
+                col_ind.append(bottom_node)
+                data.append(weight_bottom)
+
+            if i < rows - 1 and j < cols - 1:
+                bottom_right_node = to_node_index(i + 1, j + 1, cols)
+                weight_bottom_right = mat[i + 1, j + 1]
+                row_ind.append(node)
+                col_ind.append(bottom_right_node)
+                data.append(weight_bottom_right)
+
```

Comment:
why do you have to consider the three options (bottom, right, bottom_right) instead of just two (bottom, right) again?

Reply:
Seems to be more precise. Imo it should be ideal to to have right and bottom_right only, but that leads to breaking searches.